File: c/w1.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>
#include <assert.h>

// Favorite Food
// Print out your favorite food
char* fav_food(void) {
    return "Crab Legs";
}

// Number Exercise
// Write 5 examples of expressions using operations
void some_operations(void) {
    printf("%d\n", 4 + 2);
    printf("%f\n", 4 + 2.2);
    printf("%f\n", 4 / 2.33);
    printf("%d\n", 5 % 2);
    printf("%f\n", 2.55 - 1.441);
}

// Boolean Exercise
// Write some boolean operations

void bools_example(void) {
    bool x = false;
    x = true && false;
    printf("%s\n", x ? "true" : "false");
    x = true || false;
    printf("%s\n", x ? "true" : "false");
    x = !false;
    printf("%s\n", x ? "true" : "false");
}

// Variables Exercise
// Declare a variable called "first_name" and "last_name" yaddy yada
void full_name(char* name) {
    strcat(name, "David ");
    strcat(name, "Rogers");
}

// Strings Exercise
// Show concatenation and find string length
// I previously showed concatenation, lets implement a string length function
int string_length(char *string) {
    int length = 0;
    while (*string++ != '\0') length++;
    return length;
}

// Goodbye (*)
// say goodbye to given string
char *goodbye(char *name) {
    static char goodbye[100] = "Goodbye, ";
    strcat(goodbye, name);
    return goodbye;
}

// Plus Five
// Write a function that adds 5
int plus_five(int num) {
    return num + 5;
}

// Whisper
// Write a function that returns whispered version
char* whisper(char *string) {
    static char whisper[255] = "...";
    char *ellipse = "...";
    char *string_p = string;
    while (*string != '\0') {
        *string = tolower(*string);
        string++;
    }
    return strcat(whisper, strcat(string_p, ellipse));
}

// Averager
// Function which returns average of numbers
float average(float *nums, int total) {
    int count = total;
    float acc = 0;
    while (count-- >= 0) acc += nums[count];
    return acc / total;
}

// Is Odd
// Determine if number is odd
int is_odd(int n) {
    if (n <= 0) return 0;
    return n % 2 == 0 ? 0 : 1;
}

// Log Between
// Print every numbers between a low and high value
// NOTE: First we need to implement a clean way to display arrays
void print_arr(int *arr, int length) {
    printf("{");
    for (int i = 0; i < length; i++) {
        if (i == length - 1) {
            printf("%d", arr[i]);
            break;
        }
        printf("%d, ", arr[i]);
    }
    printf("}\n");
}

// Dynamic array implementation, not this only works for positive a and b, b > a
int *new_array(int a, int b) {
    size_t length = abs(b - a);
    int *p_arr;
    p_arr = (int*) malloc(length * sizeof(int));
    int *p_arr_begin = p_arr;
    if (p_arr == NULL) {
        printf("\nERROR: Allocation of memory failed in new_array().\n");
        exit(-1);
    }
    for (size_t i = 0; i < length; i++) {
        *p_arr = a++;
        p_arr++;
    }
    return p_arr_begin;
}

// Print Fives
// Print multiples of five in a range
// Simply generalize new_array()
int *new_mult_array(int a, int b, int mult) {
    size_t length = abs(b - a) / mult;
    int *p_arr;
    p_arr = (int*) malloc(length * sizeof(int));
    int *p_arr_begin = p_arr;
    if (p_arr == NULL) {
        printf("\nERROR: Allocation of memory failed in new_array().\n");
        exit(-1);
    }
    for (int i = a; i <= b; i++) {
        if (a % mult == 0) {
            *p_arr = a;
            p_arr++;
        }
        a++;
    }
    return p_arr_begin;
}

// Add Array
// Write a function which takes two arrays and returns the two combined into one.
int* combined_array(int* arr1, int* arr2, size_t length1, size_t length2) {
    int* combined_array = (int*) malloc(sizeof(int)*(length1+length2));
    for (size_t i = 0; i < length1; i++)
        combined_array[i] = arr1[i];
    for (size_t i = 0; i < length2; i++)
        combined_array[length1+i] = arr2[i];
    return combined_array;
}

// Vowel Counter
bool is_vowel(char c)  {
    switch (c) {
        case 'a':
        case 'e':
        case 'i':
        case 'o':
        case 'u': return true;
        default: return false;
    }
}

int vowel_counter(char* string) {
    int count = 0;
    while (*string++ != '\0') if (is_vowel(tolower(*string))) count++;
    return count;
}

bool is_substring(char *sub, char *string) {
    char* ret;
    ret = strstr(string, sub);
    if (ret)
        return true;
    else
        return false;
}

char** fizz_buzz(size_t max) {
    // Max string length is of "Fizz Buzz" -- 10 chars
    char** fizz_arr = malloc(max * sizeof(char*));
    for (size_t i = 0; i < max; i++)
        fizz_arr[i] = (char*) malloc(10 * sizeof(char));

    for (int i = 1; i < (int) max; i++) {
        if (!(i%3) && !(i%5)) strcpy(fizz_arr[i], "Fizz Buzz");
        else if (!(i%3)) strcpy(fizz_arr[i], "Fizz");
        else if (!(i%5)) strcpy(fizz_arr[i], "Buzz");
        else sprintf(fizz_arr[i], "%d", i);
        printf("%s\n", fizz_arr[i]);
    }
    return fizz_arr;
}

char* pig_word(char *string) {
    size_t length = (size_t) string_length(string);
    char* pig_string;
    int count = 0;
    char first = string[0];
    char cfirst;
    char* cons_buffer = malloc(10*sizeof(char));

    if(strstr("aeiou", &first)) {
        pig_string = malloc((length+4)*sizeof(char));
        strcpy(pig_string, string);
        strcat(pig_string, "yay");
    } else {
        pig_string = malloc((length+3)*sizeof(char));
        while (first != '\0') {
            if(strstr("aeiou", &first)) break;
            strcat(cons_buffer, &first);
            first = string[++count];
        }
        while (first != '\0') {
            strcat(pig_string, &first);
            first = string[++count];
        }
        count = 0;
        cfirst = cons_buffer[count];
        while (cons_buffer[count] != '\0') {
            strcat(pig_string, &cfirst);
            cfirst = cons_buffer[++count];
        }
        strcat(pig_string, "ay");
    }
    free(cons_buffer);
    return pig_string;
}

void pig_latin(char **string, int len) {
    //NOTE: look into strtok
}

bool is_prime(int n) {
    for (int i = 2; i < n; i++) {
        if (!(n % i)) return false;
    }
    return true;
}

int main(int argc, char **argv) {
    int start_num = atoi(argv[2]);
    printf("%s\n", fav_food());
    some_operations();
    bools_example();

    // First fix for referencing memory in stack, can also use static
    // inside of function call, look at goodbye();
    char final_name[100];
    full_name(final_name);
    printf("%s\n", final_name);

    // We'll take this from command line, remember argv[1] is string
    printf("%d\n", string_length(argv[1]));
    printf("%s\n", goodbye(argv[1]));

    printf("%d\n", plus_five(atoi(argv[2])));
    printf("%s\n", whisper(argv[1]));

    float number_set[5] = {1,2,3,4,5};
    printf("Average is: %.2f\n", average(number_set, 5));

    printf("Is odd: %s\n", is_odd(start_num) == 1 ? "true" : "false");

    int start = 1;
    int end = 21;
    int *test_arr = new_array(start,end);
    int *test_mult_arr = new_mult_array(start,end, 5);
    print_arr(test_arr, end-start);
    print_arr(test_mult_arr, (end-start)/5);
    free(test_arr);
    free(test_mult_arr);

    int aT1[3] = {1,2,3};
    int aT2[4] = {4,5,6,7};
    int* new_arr = combined_array(aT1,aT2,3,4);
    print_arr(new_arr,7);
    free(new_arr);

    int vowel_count = vowel_counter(argv[1]);
    printf("VOWEL COUNT: %d\n", vowel_count);

    printf("Is \"%s\" a substring of \"%s\"? %s\n", "Dog", "Cat ate a thing",
           is_substring("Dog", "Cat ate a thing") ? "yes" : "no");

    printf("Is \"%s\" a substring of \"%s\"? %s\n", "Cat", "Cat ate a thing",
           is_substring("Cat", "Cat ate a thing") ? "yes" : "no");

    char** fizz_test = fizz_buzz(16);
    free(fizz_test);

    char* old_string = "sham";
    char* piggy = pig_word(old_string);
    printf("\nPIGGY STRING: %s\n", piggy);
    free(piggy);

    int test_prime = is_prime(6);
    printf("Is Prime? %s\n", test_prime ? "true" : "false");

    return 0;
}

File: haskell/w1.hs
import System.Posix (SystemID(version))
import Data.Char (toLower)
import Data.List (isInfixOf, isPrefixOf, isSuffixOf)

-- Favorite Food
-- Print out your favorite food
favFood = putStrLn "Crab Legs"

-- Number Exercise
-- Write 5 examples of expressions using operations
someOperations = do
  print $ 4 + 2.5
  print $ 4 * 2
  print $ 4 / 2
  print $ 4 / 2.4
  print $ 5 `mod` 2
  print $ 2.3 - 1.333

-- Boolean Exercise
-- Write some examples of expressions using logic operations
boolsExample = do
  print $ True == True
  print $ True == False
  print True
  print $ True && False
  print $ True || False
  print $ True && True

-- Variables Exercise
-- Declare a cariable called "firstName" and "lastName" yaddy yadda
firstName :: String
firstName = "David"
lastName :: String
lastName = "Rogers"
fullName :: String
fullName = concat [firstName, " ", lastName]
varsExample = do
  print $ concat [firstName, " ", lastName]
  print fullName

-- Strings Exercise
-- Show concatenation and find string length
newString :: String
newString = concat ["Hello,", " ", "World"]
stringLength :: Int
stringLength = length newString
calculateLength = print stringLength

-- Goodbye (*)
-- Write a function that takes a string and returns a string saying
-- bye to that name
goodbye :: String -> String
goodbye s = "Goodbye, " ++ s

-- Plus Five
-- Write a function plusFive that takes in a number as an argument
-- and returns the sum + 5.
plusFive :: Int -> Int
plusFive x = x + 5


-- Whisper
-- Write a function that takes a string and returns whispered version
-- We map toLower over string
whisper :: String -> String
whisper s = "..." ++ map toLower s ++ "..."

-- Averager
-- Function which returns average of list of nums
averager :: [Float] -> Float
averager xs = sum xs / fromIntegral (length xs);

-- Is Odd
-- Function determines if number is odd or not
isOdd :: Int -> Bool
isOdd n
  | odd n = True
  | otherwise = False

-- Log Between
-- Print every number between a low and high value
logBetween :: Int -> Int -> [Int]
logBetween a b = [a..b]

-- Print Fives
-- Function to print out multiple of fives in range
printFives :: Int -> Int -> [Int]
printFives a b = [x | x <- [a..b], x `mod` 5 == 0]

-- Add Arrays
-- Write function that returns two combined arrays
combineArrays :: [Int] -> [Int] -> [Int]
combineArrays as bs = as ++ bs

-- Doubler
doubleList :: [Int] -> [Int]
doubleList xs = [2*x | x <- xs]

-- Vowel Counter
isVowel :: Char -> Bool
isVowel c = c `elem` "aeiou"

vowelCounter :: String -> Int
vowelCounter s = length $ filter isVowel s

-- Testing IO monad
promptCountVowels :: IO ()
promptCountVowels = do
  putStrLn "Enter a string to count vowels: "
  s <- getLine
  print (vowelCounter s)

-- Is Substring
-- Check if given string is substring of other
isSubstring :: String -> String -> Bool
isSubstring = isInfixOf

-- Fizz Buzz
fizzBuzzOut :: Int -> String
fizzBuzzOut k
  | k `mod` 5 == 0 && k `mod` 3 == 0 = "Fizz Buzz"
  | k `mod` 3 == 0 = "Fizz"
  | k `mod` 5 == 0 = "Buzz"
  | otherwise = show k

fizzBuzz :: Int -> [String]
fizzBuzz max = map fizzBuzzOut [1..max]

-- Pig Latin
-- Takes a string and out puts it in pig latin
-- NOTE: Remember that using head/tail here would be UNSAFE
startsVowel :: String -> Bool
startsVowel (c:_)
  | c `elem` "aeiou" = True
  | otherwise = False

word2Pig :: String -> String
word2Pig w
  | startsVowel w = w ++ "yay"
  | otherwise = go w where
      go (c:cs)
        | not $ startsVowel (c:cs) = go (cs ++ [c])
        | otherwise = (c:cs) ++ "ay"

toPig :: String -> String
toPig = unwords . map word2Pig . words

-- Pig Latin (break/span)
-- Implementation using break/span
-- NOTE: note that (pre,post) is essentially destructuring
word2PigAlt :: String -> String
word2PigAlt w
  | startsVowel w = post ++ "yay"
  | otherwise = post ++ pre ++ "ay" where
      (pre,post) = break isVowel w

toPigAlt :: String -> String
toPigAlt = unwords . map word2PigAlt . words

-- Is Prime
isPrime :: Int -> Bool
isPrime n
  | n < 1 = False
  | n == 1 = True
  | sum [m | m <- [2..(n-1)], n `mod` m == 0] == 0 = True
  | otherwise = False

-- reverseString
_reverseString :: String -> String
_reverseString = reverse

-- Reverse String, actual implementation
reverseString :: String -> String
reverseString (c:cs)
  | null cs = [c]
  | otherwise = reverseString cs ++ [c]


-- Range write a function that returns a range. lel
returnRange :: Int -> Int -> [Int]
returnRange a b = [a..b]
